src/main/java/edu/rpi/legup/model/rules/CaseRule.java

```diff
@@ -125,6 +125,31 @@ public String checkRuleAt(TreeTransition transition, PuzzleElement puzzleElement
      */
     @Override
     public abstract String checkRuleRawAt(TreeTransition transition, PuzzleElement puzzleElement);
+
+    /**
+     * Returns the elements necessary for the cases returned by getCases(board,puzzleElement) to be valid
+     * Overridden by case rules dependent on more than just the modified data
+     *
+     * @param board         board state at application
+     * @param puzzleElement selected puzzleElement
+     * @return List of puzzle elements (typically cells) this application of the case rule depends upon.
+     * Defaults to any element modified by any case
+     */
+    public List<PuzzleElement> dependentElements(Board board, PuzzleElement puzzleElement) {
+        List<PuzzleElement> elements = new ArrayList<>();
+
+        List<Board> cases = getCases(board,puzzleElement);
+        for (Board caseBoard : cases) {
+            Set<PuzzleElement> data = caseBoard.getModifiedData();
+            for (PuzzleElement element : data) {
+                if(!elements.contains(board.getPuzzleElement(element))){
```

Comment:
Your code still has inconsistent spacing (this comment is only the surface of formatting issues).